### PR TITLE
Wait for ScriptEngine to quit

### PR DESCRIPTION
**Pull request by zzmp**

### CODE REVIEW

**Comment by zzmp** [comment]
#### Testing

Locally, I triggered a crash by disconnecting from a domain. You may want to run an infinite loop in an entity and then do so.

**Comment by ZappoMan** [comment]
what's the goal of this change? it looks like it will still crash.

**Comment by hifi-gustavo** [comment]
Build console output for commit: 17be367d65b3321009a6b4c3e4dde784b9a8ee5b available at: https://gist.github.com/dc7a30cbaf1ca7161dbef345c30e7f4d<br/><br/>Installers available for download at: <br/>Windows: http://cachefly.highfidelity.com/HighFidelity-Beta-PR7863.exe<br/>OS X: http://cachefly.highfidelity.com/HighFidelity-Beta-PR7863.dmg

File: libraries/script-engine/src/ScriptEngine.cpp [hunk 1]
```diff
@@ -309,13 +310,18 @@ void ScriptEngine::waitTillDoneRunning() {
...
             // If we've been waiting a second or more, then tell the script engine to stop evaluating
             static const auto MAX_SCRIPT_EVALUATION_TIME =  USECS_PER_SECOND;
+            static const auto MAX_SCRIPT_QUITTING_TIME = 1.5 * MAX_SCRIPT_EVALUATION_TIME;
             auto elapsedUsecs = usecTimestampNow() - startedWaiting;
             if (elapsedUsecs > MAX_SCRIPT_EVALUATION_TIME) {
-                qCDebug(scriptengine) <<
-                    "Script " << scriptName << " has been running too long [" << elapsedUsecs << " usecs] quitting.";
-                abortEvaluation(); // to allow the thread to quit
-                workerThread->quit();
-                break;
```
Comment (zzmp) [inline libraries/script-engine/src/ScriptEngine.cpp]: Breaking early allowed the destruction of the ScriptEngine while its thread was still running, causing a crash.

File: libraries/script-engine/src/ScriptEngine.cpp [hunk 2]
```diff
@@ -309,13 +310,18 @@ void ScriptEngine::waitTillDoneRunning() {
...
+                if (!isQuitting) {
+                    qCDebug(scriptengine).nospace() <<
+                        "Script " << scriptName << " has been running too long [" << elapsedUsecs << "usecs] quitting.";
+                    workerThread->quit(); // avoid restarting evaluations
+                    abortEvaluation(); // break out of current evaluation
+                    isQuitting = true;
+                } else if (elapsedUsecs > MAX_SCRIPT_QUITTING_TIME) {
+                    qFatal("Script %s has been running for %dus", scriptName.toLocal8Bit(), elapsedUsecs);
```
Comment (zzmp) [inline libraries/script-engine/src/ScriptEngine.cpp]: If the thing never dies, we need to crash, but at least we can log it.

File: libraries/script-engine/src/ScriptEngine.cpp [hunk 3]
```diff
@@ -309,13 +310,18 @@ void ScriptEngine::waitTillDoneRunning() {
...
-                workerThread->quit();
-                break;
+                if (!isQuitting) {
+                    qCDebug(scriptengine).nospace() <<
+                        "Script " << scriptName << " has been running too long [" << elapsedUsecs << "usecs] quitting.";
+                    workerThread->quit(); // avoid restarting evaluations
+                    abortEvaluation(); // break out of current evaluation
+                    isQuitting = true;
+                } else if (elapsedUsecs > MAX_SCRIPT_QUITTING_TIME) {
+                    qFatal("Script %s has been running for %dus", scriptName.data(), (int)elapsedUsecs);
```
Comment (zzmp) [inline libraries/script-engine/src/ScriptEngine.cpp]: If the thing never dies, we need to crash, but at least we can log it.

**Comment by ZappoMan** [comment]
I feel like we're running faster but standing still on this one. Let's talk through the corner cases and get this right.

**Comment by zzmp** [comment]
test this please

**Comment by hifi-gustavo** [comment]
Build console output for commit: 96d842ed424e3aa4cdbedd180e0192a20ab921cb available at: https://gist.github.com/dc7a30cbaf1ca7161dbef345c30e7f4d<br/><br/>Installers available for download at: <br/>Windows: http://cachefly.highfidelity.com/HighFidelity-Beta-PR7863.exe<br/>OS X: http://cachefly.highfidelity.com/HighFidelity-Beta-PR7863.dmg

**Comment by hifi-gustavo** [comment]
Build console output for commit: 10d1bf0e58e74adffc8f96f25dd43be7e1ee67a7 available at: https://gist.github.com/dc7a30cbaf1ca7161dbef345c30e7f4d<br/><br/>Installers available for download at: <br/>Windows: http://cachefly.highfidelity.com/HighFidelity-Beta-PR7863.exe<br/>OS X: http://cachefly.highfidelity.com/HighFidelity-Beta-PR7863.dmg

File: libraries/script-engine/src/ScriptEngine.cpp [hunk 4]
```diff
@@ -309,13 +310,18 @@ void ScriptEngine::waitTillDoneRunning() {
...
-                abortEvaluation(); // to allow the thread to quit
-                workerThread->quit();
-                break;
+                if (!isQuitting) {
+                    qCDebug(scriptengine).nospace() <<
+                        "Script " << scriptName << " has been running too long [" << elapsedUsecs << "usecs] quitting.";
+                    workerThread->quit(); // avoid restarting evaluations
+                    abortEvaluation(); // break out of current evaluation
+                    isQuitting = true;
+                } else if (elapsedUsecs > MAX_SCRIPT_QUITTING_TIME) {
```
Comment (zzmp) [inline libraries/script-engine/src/ScriptEngine.cpp]: Remove `isQuitting`.

Replace with
```
workerThread->wait(MAX_SCRIPT_WAIT_TIME);
qFatal(...);
```

Comment (zzmp) [inline libraries/script-engine/src/ScriptEngine.cpp]: Use terminate instead - it *might* not crash.

**Comment by zzmp** [comment]
Things to read before updating this PR:
- http://doc.qt.io/qt-5/threads-qobject.html
- https://wiki.qt.io/QThreads_general_usage

Will post any new insights about threads to better understand this situation, and hopefully get a good repro to test.
